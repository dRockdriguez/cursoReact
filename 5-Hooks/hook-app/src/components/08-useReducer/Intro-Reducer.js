const initialState = [
  {
    id: 1,
    todo: "Aprender react",
    done: false,
  },
];

const todoReducer = (state = initialState, action) => {
  if (action?.type) {
    return [...state, action.payload];
  }

  return state;
};

let todos = todoReducer();

console.table(todos);

const newTodo = {
  id: 2,
  todo: "Aprender vue",
  done: false,
};

const action = {
  type: "ADD",
  payload: newTodo,
};

todos = todoReducer(todos, action);

console.table(todos);
