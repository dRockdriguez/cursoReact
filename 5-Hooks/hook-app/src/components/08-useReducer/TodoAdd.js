import React from "react";
import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({ handleAddTodo }) => {
  const [{ description }, handleInputChange, reset] = useForm({
    description: "",
  });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length < 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      description,
      done: false,
    };

    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Añadir tarea</h4>
      <hr />

      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="description"
          placeholder="Descripción de la tarea..."
          autoComplete="off"
          className="form-control"
          onChange={handleInputChange}
          value={description}
        />

        <div className="d-grid gap-2">
          <button
            className="btn btn-outline-primary mt-2 btn-block"
            type="submit"
          >
            Añadir
          </button>
        </div>
      </form>
    </>
  );
};
