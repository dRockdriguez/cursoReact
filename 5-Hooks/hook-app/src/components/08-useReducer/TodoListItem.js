import React from "react";

export const TodoListItem = ({ todo, index, toggleState, deleteTodo }) => {
  const { id, done, description } = todo;
  
  return (
    <>
      <li className="list-group-item" key={id}>
        <p
          className={`${done ? "complete" : ""}`}
          onClick={() => toggleState(id)}
        >
          {index + 1}. {description}
        </p>
        <button className="btn btn-danger" onClick={() => deleteTodo(id)}>
          Borrar
        </button>
      </li>
    </>
  );
};
