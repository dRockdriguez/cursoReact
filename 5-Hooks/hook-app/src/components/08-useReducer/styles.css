
body {
  padding: 50px;
}

p {
    cursor: pointer;
    margin: 0px !important;
    padding: 0px !important;
}

.complete {
    text-decoration: line-through;
}
.list-group-item {
  display: flex;
  align-items: center;
  justify-content: space-between;
  align-content: center;
}
