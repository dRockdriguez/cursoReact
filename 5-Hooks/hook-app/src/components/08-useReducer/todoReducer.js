export const todoReducer = (state = [], action) => {
  if (action) {
    const { type, payload } = action;

    switch (type) {
      case "ADD":
        return [...state, payload];
      case "DELETE":
        return state.filter((todo) => todo.id !== payload);
      default:
        return state;
    }
  }
  return [];
};
