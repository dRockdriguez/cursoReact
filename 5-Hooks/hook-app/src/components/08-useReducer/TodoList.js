import React from "react";
import { TodoListItem } from "./TodoListItem";

export const TodoList = ({ todos, toggleState, deleteTodo }) => {
  return (
    <>
      <ul className="list-group list-group-flash">
        {todos.map((todo, index) => (
          <TodoListItem
            key={todo.id}
            todo={todo}
            index={index}
            toggleState={toggleState}
            deleteTodo={deleteTodo}
          />
        ))}
      </ul>
    </>
  );
};
