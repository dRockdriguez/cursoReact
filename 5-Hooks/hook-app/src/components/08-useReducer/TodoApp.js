import React, { useEffect, useReducer } from "react";
import "./styles.css";
import { todoReducer } from "./todoReducer";
import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const handleAddTodo = (newTodo) => {
    const action = {
      type: "ADD",
      payload: newTodo,
    };

    dispatch(action);
  };

  const deleteTodo = (id) => {
    const action = {
      type: "DELETE",
      payload: id,
    };

    dispatch(action);
  };

  const toggleState = (id) => {
    const action = {
      type: "TOGGLE",
      payload: id,
    };
    dispatch(action);
  };

  return (
    <>
      <h1>
        Todo APP <small>({todos.length})</small>
      </h1>
      <hr />
      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            toggleState={toggleState}
            deleteTodo={deleteTodo}
          />
        </div>
        <div className="col-5">
          <TodoAdd handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </>
  );
};
