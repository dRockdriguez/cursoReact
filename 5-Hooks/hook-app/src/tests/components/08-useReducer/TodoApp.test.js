import { todos as initialState } from "../../fixtures/demoTodos";
import { shallow, mount } from "enzyme";
import { TodoApp } from "../../../components/08-useReducer/TodoApp";
import { act } from "@testing-library/react";


describe("TodoApp", () => {
  test("Debe mostrar el componente", () => {
    const wrapper = shallow(<TodoApp />);    
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe agregar un todo", () => {
    Storage.prototype.setItem = jest.fn(()=>{});

    const wrapper = mount(<TodoApp />);

    act(() => {
      wrapper.find("TodoAdd").prop("handleAddTodo")(initialState[0]);
    });

    expect(wrapper.find("h1").text().trim()).toBe("Todo APP (1)");
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });
});
