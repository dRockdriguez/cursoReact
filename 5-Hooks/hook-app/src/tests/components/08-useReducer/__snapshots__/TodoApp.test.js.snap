// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp Debe mostrar el componente 1`] = `
<Fragment>
  <h1>
    Todo APP 
    <small>
      (
      0
      )
    </small>
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        deleteTodo={[Function]}
        todos={Array []}
        toggleState={[Function]}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
